### ISSUE TITLE
Importing Hyperopt Results into the strategy.

### ISSUE THREAD

**Issue by Majid-Mahzarnia** [freqtrade/freqtrade#5257]
## Your question

Hello all,
I ran 4 Freqtrade bots on different machines with different coin assets and same defaults sample strategy, then I started hyperopt to optimize sample strategy and improve it,
"docker-compose run --rm freqtrade hyperopt --hyperopt SampleStrategy --hyperopt-loss OnlyProfitHyperOptLoss --strategy SampleStrategy -i 15m -e 5000 --spaces all"
After that I imported hyperopt results exactly into the default sample strategy and started bots. but there was no difference in trading results between before and after importing hyperopt results. I think I have a problem in importing results into the strategies. to clear this, I just copied and pasted Buy and Sell hyperspace params, at the end of this line:

This class is a sample. Feel free to customize it.
class SampleStrategy(IStrategy):
then I replaced ROI and Trailing stop and Stoploss values in strategies .
I'm not sure if this is the correct way, and I haven't found any step by step how to helps to make sure this way is correct. that's why I got confused.
Could you please help me to make sure if this is the correct way to import hyperopt results.
Thanks

**Comment by xmatthias** [comment]
Now that didn't improve over #5256 ... - the issue template is gone again (that's where yoou list all your versions).
That is the important part - which will tell us the version you're using - which will help tailor the answer to your particular problem.

**Comment by Majid-Mahzarnia** [comment]
OK, I see.

OS:
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal
####
Docker version 20.10.2, build 20.10.2-0ubuntu1~20.04.2
####
freqtrade 2021.6
####
Using CCXT 1.51.77 from docker image
####
python is not installed because Freqtrade is running on docker.
####
Thanks

**Comment by xmatthias** [comment]
You will need to be careful with what you're doing - i think you're mixing 2 different methods to run/configure hyperopt.

When using a explicit hyperopt file (`--hyperopt ...` - you're using the "old" method to run hyperopt.
This method requires you to adjust the code of your `populate_*` methods to match the result of hyperopt - which requires some understanding of how pandas conditions work to get right sometimes.

The relevant documentation part for this is the [legacy hyperopt](https://www.freqtrade.io/en/stable/advanced-hyperopt/#legacy-hyperopt) approach - which tries to show how to convert results into a strategy with some easy examples. 

---

The new (and recommended) method does not use an explicit hyperopt file anymore - but uses Strategy Parameters.
Everything that is configured in the strategy as `*Parameter` can be optimized - results of which are then used within the strategy.
Only in this case can you copy/paste the results into the strategy and expect changes to the result.

I'd recommend to read about this in the [hyperopt documentation](https://www.freqtrade.io/en/stable/hyperopt/), which documents the new method.

**Comment by Majid-Mahzarnia** [comment]
well, Thank you, 
because I got confused I want to make it clear to myself. please confirm this scenario:
these are my config and strategy files (I changed their extensions to .txt to attach them.)
[sample_strategy.py.txt](https://github.com/freqtrade/freqtrade/files/6796472/sample_strategy.py.txt)
[config.json.txt](https://github.com/freqtrade/freqtrade/files/6796473/config.json.txt)
Than my hypeyopt result is:
###

2021-07-10 15:32:21,856 - freqtrade.optimize.hyperopt - INFO - 5000 epochs saved to '/freqtrade/user_data/hyperopt_results/strategy_SampleStrategy_2021-07-10_09-35-46.fthypt'.

Best result:

  1425/5000:      4 trades. 2/2/0 Wins/Draws/Losses. Avg profit   0.50%. Median profit   0.00%. Total profit  1.98046463 USDT (   0.20%). Avg duration 2:04:00 min. Objective: -5846.62370


    # Buy hyperspace params:
    buy_params = {
        "mfi-value": 17,
        "fastd-value": 33,
        "adx-value": 42,
        "rsi-value": 28,
        "mfi-enabled": False,
        "fastd-enabled": False,
        "adx-enabled": True,
        "rsi-enabled": True,
        "trigger": "sar_reversal",
    }

    # Sell hyperspace params:
    sell_params = {
        "sell-mfi-value": 85,
        "sell-fastd-value": 64,
        "sell-adx-value": 81,
        "sell-rsi-value": 66,
        "sell-mfi-enabled": True,
        "sell-fastd-enabled": False,
        "sell-adx-enabled": False,
        "sell-rsi-enabled": True,
        "sell-trigger": "sell-sar_reversal",
    }

    # ROI table:
    minimal_roi = {
        "0": 0.088,
        "17": 0.078,
        "47": 0.02,
        "137": 0
    }

    # Stoploss:
    stoploss = -0.044

    # Trailing stop:
    trailing_stop = True
    trailing_stop_positive = 0.193
    trailing_stop_positive_offset = 0.28
    trailing_only_offset_is_reached = True
###

1- Then I will edit the strategy file and paste above text into it like this "at the end of class line":

####
# This class is a sample. Feel free to customize it.
class SampleStrategy(IStrategy):

    # Buy hyperspace params:
    buy_params = {
        "mfi-value": 17,
        "fastd-value": 33,
        "adx-value": 42,
        "rsi-value": 28,
        "mfi-enabled": False,
        "fastd-enabled": False,
        "adx-enabled": True,
        "rsi-enabled": True,
        "trigger": "sar_reversal",
    }

    # Sell hyperspace params:
    sell_params = {
        "sell-mfi-value": 85,
        "sell-fastd-value": 64,
        "sell-adx-value": 81,
        "sell-rsi-value": 66,
        "sell-mfi-enabled": True,
        "sell-fastd-enabled": False,
        "sell-adx-enabled": False,
        "sell-rsi-enabled": True,
        "sell-trigger": "sell-sar_reversal",
    }

    """
    This is a sample strategy to inspire you.
    More information in https://www.freqtrade.io/en/latest/strategy-customization/

    You can:
        :return: a Dataframe with all mandatory indicators for the strategies
    - Rename the class name (Do not forget to update class_name)
    - Add any methods you want to build your strategy
    - Add any lib you need to build your strategy

    You must keep:
    - the lib in the section "Do not remove these libs"
    - the methods: populate_indicators, populate_buy_trend, populate_sell_trend
    You should keep:
    - timeframe, minimal_roi, stoploss, trailing_*
    """
    # Strategy interface version - allow new iterations of the strategy interface.
    # Check the documentation or the Sample strategy to get the latest version.
    INTERFACE_VERSION = 2

    # Minimal ROI designed for the strategy.
    # This attribute will be overridden if the config file contains "minimal_roi".
    minimal_roi = {
        "0": 0.088,
        "17": 0.078,
        "47": 0.02,
        "137": 0
    }

    # Optimal stoploss designed for the strategy.
    # This attribute will be overridden if the config file contains "stoploss".
     stoploss = -0.044

    # Trailing stoploss
    trailing_stop = True
    trailing_stop_positive = 0.193
    trailing_stop_positive_offset = 0.28
    trailing_only_offset_is_reached = True
    # trailing_only_offset_is_reached = False
    # trailing_stop_positive = 0.01
    # trailing_stop_positive_offset = 0.0  # Disabled / not configured
    ###
    
2- Then I restart the bot (remove database file "tradesv3.sqlite" and re run docker-compose)
    
Please let me know if it is the correct way for this specific scenario or I have to make some other changes in strategy and config file? if I need to change something else please let me know what change should I make exactly and in which part of which file.
    
Thanks a lot

**Comment by xmatthias** [comment]
as mentioned above - it will depend.
did you run hyperopt with `--hyperopt <somehyperoptclass>` - or without it (using the hyperoptable parameters from the strategy)?

**Comment by Majid-Mahzarnia** [comment]
I used following line:
"docker-compose run --rm freqtrade hyperopt --hyperopt SampleStrategy --hyperopt-loss OnlyProfitHyperOptLoss --strategy SampleStrategy -i 15m -e 5000 --spaces all"

**Comment by xmatthias** [comment]
as you include ` --hyperopt SampleStrategy ` - you'll have to follow the [legacy hyperopt](https://www.freqtrade.io/en/stable/advanced-hyperopt/#legacy-hyperopt) documentation.

As such - you'll not be able to simply copy/paste the results - but you'll have to apply them correctly to your populate* methods.

**Comment by Majid-Mahzarnia** [comment]
OK,  you mean if I don't use "--hyperopt SampleStrategy" then I can copy and paste the exact result directly into the strategy without any changes. is the following command line correct for that purpose?
"docker-compose run --rm freqtrade hyperopt  --hyperopt-loss OnlyProfitHyperOptLoss --strategy SampleStrategy -i 15m -e 5000 --spaces all"
Thanks

**Comment by xmatthias** [comment]
Please, take the effort to read and understand the documentation.

There is no "definitive" answer to this - as it will depend how your strategy is structured, and if it does use parameters or not.
